src/lib.rs

```diff
@@ -56,10 +56,41 @@ fn camera_zoom(
             let old_scale = proj.scale;
             proj.scale = (proj.scale * (1. + -scroll * 0.001)).max(cam.min_scale);
 
+            // Apply max scale constraint
             if let Some(max_scale) = cam.max_scale {
                 proj.scale = proj.scale.min(max_scale);
             }
 
+            // If there is both a min and max x, that limits how far we can zoom. Make sure we don't exceed that
+            if let (Some(min_global_x), Some(max_global_x)) = (cam.min_global_x, cam.max_global_x) {
+                // Arkessler: I was not smart enough to solve this on my own, I plugged the combination of the scaling
+                // equations into wolfram alpha
+                //
+                // scaling.x = (window.width() / (projection.right - projection.left)) * projection.scale;
+                // min_cam_x = min_global_x + ((window.width() / 2.) * scaling.x);
+                // max_cam_x = max_global_x - ((window.width() / 2.) * scaling.x);
+                //
+                // Goal: Solve for min_cam_x == max_cam_x (the point at which there's no valid camera movement)
+                //
+                // Wolfram input:
+                // using min_global_x = i, max_global_x = j, proj.r = r, proj.l = l, window.width = w, proj.scale = s
+                //
+                // solve for s (i + ((w / 2) * ((w / (r - l) * s)))) = (j - ((w / 2) * ((w / (r - l) * s))))
+                let max_global_x_scale = (((max_global_x - min_global_x)
+                    * (proj.left - proj.right))
+                    / window.width().powi(2))
```

Comment:
It feels like this would be more readable if arranged like this:

```rust
let cam_projection_width = proj.right - proj.left;
let max_window_width = max_global_x - min_global_x;
let max_global_x_scale = max_window_width / window.width() * cam_projection_width / window.width();
```

Still not sure I completely understand it, though.